Updated SLEP (#1)
Micky774/enhancement_proposals: slep023/proposal.rst

.. _slep_020:

==============================
SLEP Template and Instructions
==============================

:Author: Meekail Zain
:Status: Draft
:Type: Standards Track
:Created: 2023-9-16

Abstract
--------

This SLEP proposes the inclusion of C++ based SIMD acceleration.
The initial scope is limited to accelerating ``DistanceMetric`` implementations,
however there may be other areas which could benefit from similar acceleration.

Detailed description
--------------------

For many estimators (see `Implementation`_), a dominant bottleneck for
computation over many samples is distance computation. Currently, any distance
calculated using the ``DistanceMetric`` back-end is completed through repeated
calls to ``DistanceMetric.dist`` which operates on a pair of sample vectors. These

> distance calculations can be understood as a series of parallel arithmetic

Reviewer: Less wordy:

```suggestion
distance calculations is a series of parallel arithmetic
```

translates to a **2.5x-4x speed up**.

While SIMD may be leveraged anywhere that there are data-parallel scalar operations
operating on contiguous memory, this SLEP focuses on the use case of accelerating
``DistanceMetric`` implementations.

SIMD optimization relies in CPU-specific features, which may not be present at runtime
if compiled and run on different machines. This increases the complexity of ensuring
portability of code, however modern libraries such as xsimd (CITE) and highway (CITE)
help mitigate this complexity by providing helpful abstractions with allow for writing
architecture-agnostic implementations.

The adoption of SIMD-accelerated computation would pose no change to user-experience
aside from reduced runtimes. It is important to note that the implementations require
C++ which may introduce some friction in terms of maintainence, however the core
computation implementations are simple, with the majority of required code being
boiler-plate. The only notable change to the build system would be the compilation
and use of a runtime library to isolate the C++ necessary for the SIMD routines. 

> This SLEP proposes to use Google Highway

Reviewer: Can you explain why Highway was chosen compared to the alternatives?

> Implementation

Reviewer: Somewhere in this section, I think it's good to link to a code example of how a metric is implemented in highway.

--------------

In order to leverage SIMD instructions, even with the aforementioned libraries, a
choice must be made regarding how to dispatch to an appropriate implementation.

> Generally, one must use either static or dynamic dispatching.

Reviewer: Which dispatching mechanism does this SLEP advocate for?

I would choose one and place the explanation about the alternative one in an Appendix.
